runtime/caml/domain.h

```diff
@@ -167,16 +171,45 @@ int caml_try_run_on_all_domains(
 */
 
 
-/* barriers */
-typedef uintnat barrier_status;
-void caml_global_barrier(void);
-barrier_status caml_global_barrier_begin(void);
-int caml_global_barrier_is_final(barrier_status);
-void caml_global_barrier_end(barrier_status);
-int caml_global_barrier_num_domains(void);
+/* Barriers */
 
-int caml_domain_terminating(caml_domain_state *);
-int caml_domain_is_terminating(void);
+/* Get the number of parties expected to arrive into the barrier, i.e. the
+   number of domains participating in the STW section. In most cases the barrier
+   is used directly from an STW callback that already has the number of
+   participating domains at hand, which should be used instead. */
+int caml_global_barrier_num_participating(void);
+
+/* Arrive at the barrier and wait for all parties */
+void caml_global_barrier(void);
+/* Arrive at the barrier and wait iff there is more than one party */
+#define Caml_maybe_global_barrier(num_participating)                    \
+  do { if ((num_participating) != 1) caml_global_barrier(); } while(0)
```

Comment:
This is definitely reasonable, there's no need to expose a global barrier function that doesn't take the number of domains, since it is always available from stw callbacks. `caml_global_barrier_num_participating` is still there if necessary, though unused.